Python 2 / 3 compatibility changes [MOVED: See PR #3856]

catleeball opened the pull request:
@walkerjoe Ed (todo: find Ed's github alias) and I made several fixes to help boto be cross-compatible with Python 2 and 3. These issues were identified while working on making [gsutil's py-six-current branch](https://github.com/GoogleCloudPlatform/gsutil/tree/py-six-current) compatible with Python 2.7 and 3.5+. This was done under the guidance of @houglum

Notably, we're using the [six](https://github.com/benjaminp/six) library to help ensure byte strings vs unicode strings work together nicely. Many of the changes we made revolve around making Python 2 and Python 3's byte vs unicode strings play together nicely, along with changes in division. The ttyprint and ttywrite methods were added to allow type agnostic string printing.

Let us know if you have any questions about these changes! :slightly_smiling_face:

Commit:
Re-introducing some of Ed's changes with slight modifications

Commit:
Adding the rest of Ed's boto changes

walkerjoe committed:
urllib changes for python 2/3 compatibility

- six import and use change
  + specific import, only importing necessity from six
  + changed from six.PY3 to PY3

walkerjoe committed:
Adapted ttywrite to use any type of writer and using replace on encoding

Previous version errored on unknown types of writers. New version
attempts to write in both unicode and binary and if both of those fail,
raises exception.

Encoding will try to use 'backslashreplace' when unicode cannot be
converted to utf-8.

walkerjoe committed:
Revised checks for get_contents_to_file()

Same checks apply to both Python 2 and 3, consolidated logic to handle
both versions.

walkerjoe committed:
Multiple imports removed for 'six' package

walkerjoe committed:
Consolidate six imports

catleeball committed:
Merge pull request #1 from walkerjoe/boto-six

six package updates

Commit:
Merge remote-tracking branch 'boto-boto/develop' into boto-six

walkerjoe commented on boto/gs/key.py:
```diff
@@ -707,7 +713,12 @@ def set_contents_from_string(self, s, headers=None, replace=True,
+                s = s.encode('utf-8')
+            fp = io.BytesIO(s)
+        else:
+            fp = io.BytesIO(get_utf8_value(s))
```
`io.BytesIO()` does not exist in Py2. We should change this to use `six.BytesIO`
Info: https://pythonhosted.org/six/#six.BytesIO

catleeball replied:
Nice catch! Updated to six.BytesIO

walkerjoe commented on boto/utils.py:
```diff
@@ -860,11 +864,15 @@ def notify(subject, body=None, html_body=None, to_string=None,
-    if not six.PY2 and isinstance(value, bytes):
-        return value
+    # if not six.PY2 and isinstance(value, bytes):
+    #     return value
```
Commented code should probably be removed.

catleeball replied:
Got it, removed

Commit:
Updated BytesIO, removed comments

Updated a call from io.BytesIO to six.BytesIO for Python 2/3
compatibility. Extranious comment removed.

catleeball commented:
Joe and I are currently investigating and fixing errors that Travis found. :construction_worker_woman:

Commit:
Removed typo from accidental paste

houglum commented on boto/gs/bucket.py:
```diff
@@ -40,14 +43,20 @@
 LIFECYCLE_ARG = 'lifecycle'
 STORAGE_CLASS_ARG='storageClass'
-ERROR_DETAILS_REGEX = re.compile(r'<Details>(?P<details>.*)</Details>')
+ERROR_DETAILS_REGEX_STR = r'<Details>(?P<details>.*)</Details>'
```
Nit: If this is only used to create `ERROR_DETAILS_REGEX` below (i.e. it shouldn't be "publicly" discoverable by other modules), consider naming it with a leading underscore, `_ERROR_DETAILS_REGEX_STR`

catleeball replied:
SGTM, done 👍

houglum commented on boto/gs/bucket.py:
```diff
@@ -353,6 +362,8 @@ def _get_xml_acl_helper(self, key_name, headers, query_args):
                     details = (('<Details>%s. Note that Full Control access'
                                 ' is required to access ACLs.</Details>') %
                                details)
+                    if six.PY3:
```
Probably worth adding a comment here mentioning *why* we do this -- because all arguments to `re.sub()` must be the same type.

catleeball replied:
SGTM, done.

catleeball commented:
Regarding the python 2.6 failure, it looks like one of boto's dependencies, [idna](https://github.com/kjd/idna) no longer supports Python 2.6 as of [this PR](https://github.com/kjd/idna/commit/2618ad17c86498d627ebb136143c2799ad48cbe3#diff-2eeaed663bd0d25b7e608891384b7298) due to 2.6's deprecation.

This error seems to occur in 2.6 for other recent PRs to Boto as well.

Boto team: just to check, does Boto still officially support deprecated Python versions including Python 2.6, 3.3, and 3.4?

Will continue investigating the other Travis errors 🔍

houglum commented on boto/gs/bucket.py:
```diff
@@ -40,14 +43,20 @@
 from boto.utils import get_utf8_value
 from boto.compat import six
 
+quote = six.moves.urllib.parse.quote
```
IIUC, most of the compatibility stuff should be taken from `boto.compat`, rather than directly from the source `boto.compat` takes its aliases from (e.g. its vendored version of `six`).  Can we change this line to:

```
from boto.compat import quote
```

catleeball replied:
Agreed, this is weird syntax, updated

houglum commented on boto/gs/key.py:
```diff
@@ -707,7 +713,12 @@ def set_contents_from_string(self, s, headers=None, replace=True,
         self.base64md5 = None
 
-        fp = StringIO(get_utf8_value(s))
+        if six.PY3:
```
Looking at `get_utf8_value`, I don't think this does anything different in Python 3 than in 2.  I think we can revert this block of changes.

houglum commented on boto/utils.py:
```diff
@@ -860,11 +864,12 @@ def notify(subject, body=None, html_body=None, to_string=None,
-    if not six.PY2 and isinstance(value, bytes):
+    if six.PY3:
+        if isinstance(value, bytes):
+            return value.decode('utf-8')
```
Why is this being decoded into unicode? That's not what this method's name indicates that it does.

Did you mean to just write `value.decode('utf-8')`, without the return statement, so that this fails if the bytes can't be encoded using the utf-8 codec?

catleeball replied:
I honestly don't know the rationale, this was pulled in from Ed's changes. Does changing line 869 to `value = value.decode('utf-8')` sound reasonable? I think you're right, the return statement looks like it was probably a typo or something.

catleeball replied:
Updated to `value = value.decode('utf-8')` per above in most recent commit.

houglum replied:
My issue with this change is that this method is called "get_utf8_value", meaning that the thing being returned should be a utf-8-encoded `bytes` object, not a unicode object.  The only reason that a `value.decode('utf-8')` line would be present would be to make sure that the bytes passed in could be decoded using UTF-8.

Given this confusion, I think it's worth also adding a comment to the docstring saying that the return type of this method is a `bytes` value.

houglum commented on boto/utils.py:
```diff
@@ -860,11 +864,12 @@ def notify(subject, body=None, html_body=None, to_string=None,
-
     if not isinstance(value, six.string_types):
-        value = six.text_type(value)
+        value = six.text_type(value).encode('utf-8')
```
This isn't necessary.  The next lines will detect that this is an instance of `text_type` and do the same encoding that's being done here.  This line can be reverted.

catleeball replied:
Sounds right to me, reverted.

houglum commented on boto/utils.py:
```diff
@@ -1096,3 +1101,81 @@ def parse_host(hostname):
+
+
+def ttyprint(*objects, **kwargs):
+  """ A Python 2&3 compatible version of the print function.
```
The indentation in this function is 2 spaces; it should be 4.

houglum commented on boto/s3/key.py:
```diff
@@ -1549,7 +1553,7 @@ def _get_file_internal(self, fp, headers=None, cb=None, num_cb=10,
         try:
             for bytes in self:
-                fp.write(bytes)
+                boto.utils.ttyprint(bytes, file=fp, end='')
```
The name on this method seems to be misleading -- we're not necessarily writing to a tty, correct? `fp` could refer to a file descriptor that isn't a terminal.  Maybe the method should be named `write_to_file_or_tty`?

catleeball replied:
Sounds good to me. What about `write_to_fd` or `write_to_file_descriptor`? A smidge shorter for the former, and file descriptor covers both, especially since `file=` should be able to take any kind of fd.

houglum replied:
Sounds good. Might even be worth revisiting the logic to see if we can break it up into something like:

```
if fd_is_tty():
  write_to_tty(data)
else:  # fd is a file
  write_to_file(data)
```

... if that's possible.  It's a bit tough to follow as-is.

catleeball replied:
Yeah, it's due for a refactor, I'll do break it up a bit :+1:

catleeball replied:
Since this is the single and only call in boto to ttyprint / ttywrite, I'm going to axe those functions from boto entirely and just do six.ensure_binary() here. The refactor to those functions I'll include in gsutil where those functions are used more widely

Commit:
Explicitly sending ascii to urllib.parse.quote()

Commit:
Marked ERROR_DETAILS_REGEX_STR private with leading underscore.

Commit:
Used idiomatic import for quote, added comment

Commit:
Removed extra return and unneeded decode

walkerjoe committed:
Update six to 1.12

Reference: https://github.com/benjaminp/six/blob/master/six.py

walkerjoe committed:
Changed python versions tested

Python 2.6 and 3.3 have been deprecated (with 3.4 soon to follow).
Changed tested versions to reflect current versions supported.

walkerjoe committed:
Changed python versions tested

Python 2.6 and 3.3 have been deprecated (with 3.4 soon to follow).
Changed tested versions to reflect current versions supported.

walkerjoe committed:
Clearing Test Errors

Most errors were due to encoding issues between Python 2 and 3. Added
six.ensure statements to maintain functionality.

walkerjoe committed:
Clearing Test Errors

Most errors were due to encoding issues between Python 2 and 3. Added
six.ensure statements to maintain functionality.

walkerjoe committed:
Added back deprecated python versions to TravisCI

walkerjoe committed:
Added back deprecated python versions to TravisCI

walkerjoe committed:
Added back py3.3 condition for installation scripts and additional comments

walkerjoe committed:
Added back py3.3 condition for installation scripts and additional comments

catleeball committed:
Merge pull request #2 from walkerjoe/boto-six

Boto six

catleeball committed:
Merge pull request #2 from walkerjoe/boto-six

Boto six

Commit:
Removed six upgrade commit

Commit `c30a12c85570106d51ccd82069e4e45c2a1a3353` was moved to a
separate PR https://github.com/boto/boto/pull/3852

catleeball commented:
Moved the commit for upgrading the six library to https://github.com/boto/boto/pull/3852

catleeball commented:
Moved travis update to https://github.com/boto/boto/pull/3853

Commit:
Moved Travis changes to PR #3853

catleeball commented:
Split utils.py changes into https://github.com/boto/boto/pull/3854

catleeball commented:
Split mock storage integ test fix into https://github.com/boto/boto/pull/3855

catleeball commented:
Migrated remaining changes to https://github.com/boto/boto/pull/3856 closing this PR
